apm-sniffer/apm-sdk-plugin/finagle-6.25.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/finagle/ClientTracingFilterInterceptor.java

```diff
@@ -0,0 +1,107 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.apm.plugin.finagle;
+
+import com.twitter.util.Future;
+import com.twitter.util.FutureEventListener;
+import org.apache.skywalking.apm.agent.core.context.ContextCarrier;
+import org.apache.skywalking.apm.agent.core.context.ContextManager;
+import org.apache.skywalking.apm.agent.core.context.trace.AbstractSpan;
+import org.apache.skywalking.apm.agent.core.context.trace.SpanLayer;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.MethodInterceptResult;
+
+import java.lang.reflect.Method;
+
+import static org.apache.skywalking.apm.network.trace.component.ComponentsDefine.FINAGLE;
+import static org.apache.skywalking.apm.plugin.finagle.ContextHolderFactory.getLocalContextHolder;
+import static org.apache.skywalking.apm.plugin.finagle.ContextHolderFactory.getMarshalledContextHolder;
+import static org.apache.skywalking.apm.plugin.finagle.FinagleCtxs.SW_SPAN;
+
+public class ClientTracingFilterInterceptor extends AbstractInterceptor {
+
+    private static class CacheObjects {
+        private ContextHolder marshlledContextHolder;
+        private ContextHolder localContextHolder;
+
+        private CacheObjects(ContextHolder marshlledContextHolder, ContextHolder localContextHolder) {
+            this.marshlledContextHolder = marshlledContextHolder;
+            this.localContextHolder = localContextHolder;
+        }
+    }
+
+    @Override
+    protected void onConstructImpl(EnhancedInstance objInst, Object[] allArguments) {
+
+    }
+
+    @Override
+    public void beforeMethodImpl(EnhancedInstance enhancedInstance, Method method, Object[] objects, Class<?>[] classes,
+                                 MethodInterceptResult methodInterceptResult) throws Throwable {
+        ContextCarrier contextCarrier = new ContextCarrier();
+        /*
+         * At this time, we can't know the operation name and peer address, so we just use placeholders here, the
+         * operation name will be filled by {@link AnnotationInterceptor$Rpc} and the peer address will be filled by
+         * {@link ClientDestTracingFilterInterceptor} later.
+         */
+        AbstractSpan finagleSpan = ContextManager.createExitSpan("pending", contextCarrier, "");
+
+        finagleSpan.setComponent(FINAGLE);
+        SpanLayer.asRPCFramework(finagleSpan);
+
+        ContextHolder marshlledContextHolder = getMarshalledContextHolder();
+        marshlledContextHolder.let(SWContextCarrier$.MODULE$, SWContextCarrier.of(contextCarrier));
+
+        ContextHolder localContextHolder = getLocalContextHolder();
+        localContextHolder.let(SW_SPAN, finagleSpan);
+
+        enhancedInstance.setSkyWalkingDynamicField(new CacheObjects(marshlledContextHolder, localContextHolder));
```

Comment:
Is the `enhancedInstance` used in only one thread only? If there is a race condition, you can't use the dynamic field.

Reply:
I am confused here, if the `dynamic field` is safe enough, why you need the `ContextHolder`?

Reply:
`dynamic field` could set any customized class, from my reading, it should be used to replace the `ContextHolder`.

Anyway, I think I need your further explanation about this.

Reply:
> Is the enhancedInstance used in only one thread only?
The  `enhancedInstance ` is used in mulit threads, I will remove usage of it.

> I am confused here, if the dynamic field is safe enough, why you need the ContextHolder?
`ContextHolder` is used to put `finagleSpan`,`SWContextCarrier` to Context, thus these two can be accessed by `ClientDestTracingFilterInterceptor`, even if `ClientDestTracingFilter` is executed in another thread.